source/common/grpc/buffered_message_ttl_manager.h

```diff
@@ -0,0 +1,66 @@
+#pragma once
+
+#include <queue>
+
+#include "envoy/event/dispatcher.h"
+
+#include "source/common/grpc/buffered_async_client.h"
+
+namespace Envoy {
+namespace Grpc {
+
+class BufferedMessageTtlManager {
+public:
+  BufferedMessageTtlManager(Event::Dispatcher& dispatcher, BufferedAsyncClientCallbacks& callbacks,
+                            std::chrono::milliseconds message_ack_timeout)
+      : dispatcher_(dispatcher), message_ack_timeout_(message_ack_timeout), callbacks_(callbacks),
+        timer_(dispatcher_.createTimer([this] { checkExpiredMessages(); })) {}
+
+  ~BufferedMessageTtlManager() { timer_->disableTimer(); }
+
+  void setDeadline(absl::flat_hash_set<uint64_t>&& ids) {
+    const auto expires_at = dispatcher_.timeSource().monotonicTime() + message_ack_timeout_;
+    deadline_.emplace(expires_at, std::move(ids));
+
+    if (!timer_->enabled()) {
+      timer_->enableTimer(message_ack_timeout_);
+    }
+  }
+
+  const std::queue<std::pair<MonotonicTime, absl::flat_hash_set<uint64_t>>>& deadline() {
+    return deadline_;
+  }
+
+private:
+  void checkExpiredMessages() {
+    const auto now = dispatcher_.timeSource().monotonicTime();
+
+    while (!deadline_.empty()) {
+      auto& it = deadline_.front();
+      if (it.first > now) {
+        break;
+      }
+      for (auto&& id : it.second) {
+        // If the retrieved message is a PendingFlush, it means that the message
+        // has timed out. A timeout is treated as an error, and the callback will
+        // re-buffer the message.
+        callbacks_.onError(id);
```

Comment:
I wonder if instead of reusing `BufferedAsyncClientCallbacks` this should have its own simpler callback interface? I think that might be a bit clearer? WDYT?

Reply:
Does it mean that `BufferedMessageTtlManager` exposes some callbacks and is accessed by `BufferedAsyncClient`?

Reply:
I was thinking we could just have a single `std::function<void(uint64_t)> expiry_callback` function passed to the constructor, that way we don't have any real coupling between this and the buffered async client, this just becomes a "pure" TTL management class

Reply:
fixed.